odoo/models.py

```diff
@@ -4295,43 +4290,42 @@ def _filter_access(self, operation: str):
 
         """
         if self and not self.env.su and (forbidden := self._check_access(operation)):
-            # invalidate forbidden records to avoid cache "pollution"
-            self.browse(forbidden).invalidate_recordset()
-            return self.browse(id_ for id_ in self._ids if id_ not in forbidden)
+            forbidden.invalidate_recordset()
+            return self - forbidden
         return self
 
-    def _check_access(self, operation: str) -> dict:
-        """ Return a dictionary mapping record ids to a callable that generates
-        an exception. The dictionary keys correspond to the records in ``self``
-        on which the current is not allowed to perform ``operation``. When the
+    def _check_access(self, operation: str, raise_exception: bool = False) -> Self:
```

Comment:
I liked it better before. I really don't like the fact of returning "some arbitrary record". I find other functions not much simpler and now there is a boolean argument to check. I think this function tries to do too much.
The name of the function does not match with what is returns anymore: "_filter_forbidden_access"?

Small note: in previous implementation if there is no access to the model, you could always return `{None: ...}` insteado of `self._ids or (None,)`. This way you can distinguish why you don't have access.